_drafts/code-properties-and-language-features.md

```diff
@@ -0,0 +1,50 @@
+
+Then I asked about [the `Plug.Conn` struct][conn] in the Phoenix framework, which is passed through much of the app. I asked about whether this made the entire application fragile, because all of it required intimate knowledge of the Connection. I learned that there was a [`private`][conn-private] field on the struct, and although you aren't prevented from accessing it directly, the convention is that you shouldn't because the implementation in it can change without warning. This reminded me of some Rubyists' point that, because you can get around the private keyword, it's effectively just a convention as well.
+
+After all this, I still had a concern with [Elixir's pattern matching][pattern-matching]. It's a lot like a switch statement, and [switch statements are a code smell][switch] in OO design. It makes it difficult to change your code because switch statements proliferate: every time you have to add a new case, all the switch statements need to be updated. Instead it's best to use polymorphism to handle variations, so that you can always add a new type of object and implement the new case's behavior on it. Each place you previously had a switch statement, you now send a message instead. But Elixir has a potential answer to this as well: [protocols][protocols] allow you to define a function signature, and then you can implement it for specific data types.
+
+I had a similar question about Redux. Having a single reducer you call with actions seemed to combine all the logic in one place. But it's conventional to [compose your reducer][reducer-composition] out of smaller reducers that produce the values for individual keys in your state. So each smaller reducer only has responsibility for one piece of state, and it can just return the data unchanged for any actions that don't pertain to it. Yes, there is just one big list of action names that are strings, but if you like you can introduce namespacing into your action names to group them by data type.
+
+At this point I was really confused. Were there *any* differences in capabilities between OO and FP? Had OO's grouping of data together with functionality really accomplished anything?
+
+# Code Properties
+
+**I think it all comes down to what properties you want your code to have, and whether a given language and paradigm make it easier or more difficult.**
+
+For example, concurrency can be a big benefit for systems that need to handle a high volume of data throughput or that are making multiple asynchronous I/O requests. You can get a little concurrency in a traditional OO language like Ruby or Java, but it's incredibly hard to use it extensively in a reliable way. As my former coworker John Gallager said about concurrency on iOS in particular, "we can safely say that all nontrivial iOS applications have at least one concurrency bug." By contrast, Elixir, Rust, and JavaScript each have a different concurrency model that makes safe concurrency much more achievable.
```

Comment:
I feel like you're arguing that there is something fundamental about OO languages that makes it hard to use concurrency. Is that true? If so, what is it? (And what about Javascript's prototype-based object system?)